### PR TITLE
K8SPS-56: Autotune MySQL parameters using container resources

### PR DESCRIPTION
**Pull request by pooknull**
https://jira.percona.com/browse/K8SPS-56

Add `auto-config` configmap if there is `resources` field in `mysql` section. If a user specifies some of the lines that `auto-config` should use, they won't be included in the `auto-config`

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/19059b9/users.log) |
| We run 9 out of 9|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/19059b9b5fb81b94c642a29a4709c8f78bed9548
image: `perconalab/percona-server-mysql-operator:PR-74-19059b9`

**Review by tplavcic — CHANGES_REQUESTED** [review]
We have an issue with the PXC operator here: https://jira.percona.com/browse/K8SPXC-1028
So please wait with the fix merge so that we can update the PS operator code as well.

**Comment by pooknull** [comment]
> @pooknull please apply the changes in [percona/percona-xtradb-cluster-operator#1166](https://github.com/percona/percona-xtradb-cluster-operator/pull/1166)

@egegunes I don't see much difference

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1f7fd01/users.log) |
| We run 9 out of 9|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/1f7fd01e188ab83c86eae3ab758d45f4cd993d43
image: `perconalab/percona-server-mysql-operator:PR-74-1f7fd01`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/auto-config.log) |
| config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/scaling.log) |
| sidecars | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/users.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/0df6802/demand-backup.log) |
| We run 10 out of 10|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/0df68028a7b5c21a457a2d11059ca4bcf4513bd0
image: `perconalab/percona-server-mysql-operator:PR-74-0df6802`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/auto-config.log) |
| config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/scaling.log) |
| sidecars | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/users.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f4e864e/demand-backup.log) |
| We run 10 out of 10|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/f4e864e0297fe823df705c050bc049bb6703e9ac
image: `perconalab/percona-server-mysql-operator:PR-74-f4e864e`

File: pkg/mysql/mysql.go [hunk 1]
```diff
@@ -89,20 +89,25 @@ func ConfigMapName(cr *apiv1alpha1.PerconaServerMySQL) string {
...
-func StatefulSet(cr *apiv1alpha1.PerconaServerMySQL, initImage, configHash string) *appsv1.StatefulSet {
+func StatefulSet(cr *apiv1alpha1.PerconaServerMySQL, initImage, configHash, autoConfigHash string) *appsv1.StatefulSet {
 	labels := MatchLabels(cr)
 	spec := cr.MySQLSpec()
 	replicas := spec.Size
 	t := true
 
 	annotations := make(map[string]string)
 	annotations["percona.com/configuration-hash"] = configHash
+	annotations["percona.com/auto-configuration-hash"] = autoConfigHash
```
Comment (egegunes) [inline pkg/mysql/mysql.go:110]: why do we need it? as far as I see only thing that affects auto config is resources and if resources are updated the statefulset will be restarted anyway

File: controllers/perconaservermysql_controller.go [hunk 2]
```diff
@@ -586,6 +652,59 @@ func (r *PerconaServerMySQLReconciler) reconcileMySQLConfiguration(
 	return fmt.Sprintf("%x", md5.Sum(data)), nil
 }
 
+func getAutoTuneParams(cr *apiv1alpha1.PerconaServerMySQL, q *resource.Quantity) (string, error) {
```
[the remark this answers is not in the document]
Comment (pooknull) [reply controllers/perconaservermysql_controller.go]: d8b01c839273bef45471d8e50562a8adc387366a

File: pkg/mysql/mysql.go (hunk 1 above)
Comment (pooknull) [reply pkg/mysql/mysql.go]: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/07fa2f2ed36f7977484ff31f7bf8869e15c2ac00

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/scaling.log) |
| sidecars | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/users.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/1b89695/demand-backup.log) |
| We run 10 out of 10|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/1b89695792743dbb5557b4e2017ee56dc72a6c85
image: `perconalab/percona-server-mysql-operator:PR-74-1b89695`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/users.log) |
| groupreplication | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/groupreplication.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/f212f15/demand-backup.log) |
| We run 11 out of 11|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/f212f158454403d870db202d443c3a0b2b61dcee
image: `perconalab/percona-server-mysql-operator:PR-74-f212f15`

**Review by egegunes — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/users.log) |
| groupreplication | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/groupreplication.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-74/57f5c95/demand-backup.log) |
| We run 11 out of 11|

commit: https://github.com/percona/percona-server-mysql-operator/pull/74/commits/57f5c95aeb7972c8f3a910dfcb5c2d54cfb10d64
image: `perconalab/percona-server-mysql-operator:PR-74-57f5c95`

**Review by hors — APPROVED** [review]

**Pull request merged**